lib/posix-tty/serial.c

```diff
@@ -0,0 +1,259 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <termios.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+
+#include <uk/assert.h>
+#include <uk/file.h>
+#include <uk/file/nops.h>
+#include <uk/plat/console.h>
+#include <uk/posix-fd.h>
+#include <uk/posix-serialfile.h>
+
+
+static const char SERIAL_VOLID[] = "serial_vol";
+
+/* EOT: End Of Transmission character; ^D */
+#define SERIAL_EOT 04
+
+#define SERIAL_TERMIOS_IFLAGS (IUTF8|ICRNL)
+#define SERIAL_TERMIOS_OFLAGS (0)
+#define SERIAL_TERMIOS_CFLAGS (CREAD|CS8|B38400)
+#define SERIAL_TERMIOS_LFLAGS (ICANON|ECHO)
+
+/* Linux only fills in 19 chars; bincompat might expect this */
+#define KNCCS 19
+
+/* Values taken from termios(3) from Linux man-pages 6.05 */
+static const char SERIAL_TERMIOS_CONTROL_CHARS[KNCCS] = {
+	[VDISCARD] = 017,
+	[VEOF] = SERIAL_EOT,
+	[VEOL] = 0,
+	[VEOL2] = 0,
+	[VERASE] = 0177,
+	[VINTR] = 003,
+	[VKILL] = 025,
+	[VLNEXT] = 026,
+	[VMIN] = 1,
+	[VQUIT] = 034,
+	[VREPRINT] = 022,
+	[VSTART] = 021,
+	[VSTOP] = 023,
+	[VSUSP] = 032,
+	[VSWTC] = 0,
+	[VTIME] = 0,
+	[VWERASE] = 027,
+};
+
+
+static ssize_t serial_read(const struct uk_file *f,
+			   const struct iovec *iov, int iovcnt,
+			   off_t off, long flags __unused)
+{
+	ssize_t total = 0;
+
+	UK_ASSERT(f->vol == SERIAL_VOLID);
+	if (unlikely(off != 0))
+		return -EINVAL;
+
+	if (!uk_file_poll_immediate(f, UKFD_POLLIN))
+		return 0;
+
+	for (int i = 0; i < iovcnt; i++) {
+		char *buf = iov[i].iov_base;
+		size_t len = iov[i].iov_len;
+		char *last;
+		int bytes_read;
+
+		if (unlikely(!buf && len))
+			return -EFAULT;
+
+		bytes_read = ukplat_cink(buf, len);
```

Comment:
If `ukplat_cink` blocks then indeed serial read will also block & won't ever return `-EAGAIN`. The rest of ukfile/posix-fd* code should handle this gracefully though -- it expects files to always be nonblocking, but doesn't _demand_ them to be so; it can handle always-blocking files just fine.

Where this could pose compat issues is if an application opens the serial file `O_NONBLOCK` and then _absolutely relies_ on nonblocking behavior. That might be unlikely, since `O_NONBLOCK` is merely a hint and doesn't make `open` error out if the file cannot be opened nonblockingly, but worth considering.
In that case we may need to change `ukplat_cink` to only probe the serial input and never block.

Admittedly the current implementation is suboptimal for such a probing `ukplat_cink`, as the readiness flags are set very conservatively: POLLOUT is always set, POLLIN remains set until we get an EOF => polling code would immediately return, busy-waiting on console input. We'd much rather want to signal POLLIN only when input is available.

Reply:
Okay blocking was maybe a bit imprecise. The problem is that the serial read will busy-poll on the port. So it's not really "blocking" in the classical unix I/O sense, but worse.

Reply:
Oh I see, so it's spinning/busy-waiting on the serial input. Indeed much worse :sweat_smile: .
Let's hope actual platform serial input is never performance-critical, but definitely something worth improving in the long run.

Reply:
I think it's less a performance problem and more a function problem. AFAIK it soft-breaks the quake server, because you have repeatedly type stuff on the serial to make it continue.

Reply:
By default AFAIK we have null as stdin instead of serial console, does that also break daemons? Or are there non-interactive workloads that demand the serial stdin?

Reply:
null as stdin should be fine I guess. Might be a bit surprising for interactive workloads, but I don't expect people to do that much (and that would be likely broken anyway).